File: TaggdWin/Services/TimeTaggerClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;

namespace TaggdWin.Services
{
    /// <summary>
    /// Thin client for a self-hosted TimeTagger backend. Faithful port of the
    /// Swift <c>TimeTaggerClient</c>. Auth token is sent in the <c>authtoken</c>
    /// header; endpoints live under &lt;serverURL&gt;/api/v2/.
    /// </summary>
    public sealed class TimeTaggerClient
    {
        public string ServerUrl { get; }
        public string Token { get; }

        // One shared HttpClient for the whole process (recommended pattern).
        private static readonly HttpClient Http = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(15)
        };

        public TimeTaggerClient(string serverUrl, string token)
        {
            ServerUrl = serverUrl ?? "";
            Token = token ?? "";
        }

        public sealed class Record
        {
            [JsonPropertyName("key")] public string Key { get; set; } = "";
            [JsonPropertyName("t1")] public long T1 { get; set; }
            [JsonPropertyName("t2")] public long T2 { get; set; }
            [JsonPropertyName("mt")] public long Mt { get; set; }
            [JsonPropertyName("ds")] public string Ds { get; set; } = "";
        }

        // ---- Result kinds ----

        public enum ResultKind { Success, Unauthorized, BadUrl, Rejected, Failure }

        public readonly struct ConnectionResult
        {
            public ResultKind Kind { get; init; }
            public string Message { get; init; }
            public static ConnectionResult Ok() => new() { Kind = ResultKind.Success, Message = "" };
            public static ConnectionResult Unauthorized() => new() { Kind = ResultKind.Unauthorized, Message = "" };
            public static ConnectionResult BadUrl() => new() { Kind = ResultKind.BadUrl, Message = "" };
            public static ConnectionResult Fail(string m) => new() { Kind = ResultKind.Failure, Message = m };
        }

        public readonly struct PushResult
        {
            public ResultKind Kind { get; init; }
            public string Message { get; init; }
            public static PushResult Ok() => new() { Kind = ResultKind.Success, Message = "" };
            public static PushResult Unauthorized() => new() { Kind = ResultKind.Unauthorized, Message = "" };
            public static PushResult BadUrl() => new() { Kind = ResultKind.BadUrl, Message = "" };
            public static PushResult Reject(string m) => new() { Kind = ResultKind.Rejected, Message = m };
            public static PushResult Fail(string m) => new() { Kind = ResultKind.Failure, Message = m };
        }

        /// <summary>Builds &lt;server&gt;/api/v2/&lt;endpoint&gt;, tolerating a trailing slash.</summary>
        private Uri? ApiUrl(string endpoint)
        {
            var baseUrl = (ServerUrl ?? "").Trim();
            if (string.IsNullOrEmpty(baseUrl)) return null;
            baseUrl = baseUrl.TrimEnd('/');
            return Uri.TryCreate(baseUrl + "/api/v2/" + endpoint, UriKind.Absolute, out var uri) ? uri : null;
        }

        /// <summary>PUTs one or more records; server replies {accepted, failed, errors}.</summary>
        public async Task<PushResult> PushRecordsAsync(IReadOnlyList<Record> records)
        {
            var url = ApiUrl("records");
            if (url == null) return PushResult.BadUrl();

            string body;
            try { body = JsonSerializer.Serialize(records); }
            catch { return PushResult.Fail("Could not encode record"); }

            using var request = new HttpRequestMessage(HttpMethod.Put, url);
            request.Headers.TryAddWithoutValidation("authtoken", Token);
            request.Content = new StringContent(body, Encoding.UTF8, "application/json");

            try
            {
                using var response = await Http.SendAsync(request).ConfigureAwait(false);
                var data = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
                switch ((int)response.StatusCode)
                {
                    case 200:
                        try
                        {
                            using var doc = JsonDocument.Parse(data);
                            var root = doc.RootElement;
                            var failedCount = root.TryGetProperty("failed", out var failed) && failed.ValueKind == JsonValueKind.Array
                                ? failed.GetArrayLength() : 0;
                            if (failedCount == 0) return PushResult.Ok();
                            string firstError = "Server rejected the record";
                            if (root.TryGetProperty("errors", out var errs) && errs.ValueKind == JsonValueKind.Array && errs.GetArrayLength() > 0)
                                firstError = errs[0].GetString() ?? firstError;
                            return PushResult.Reject(firstError);
                        }
                        catch { return PushResult.Ok(); }
                    case 401:
                    case 403:
                        return PushResult.Unauthorized();
                    default:
                        return PushResult.Fail($"HTTP {(int)response.StatusCode}");
                }
            }
            catch (Exception ex)
            {
                return PushResult.Fail(ex.Message);
            }
        }

        /// <summary>Validates URL + token with a minimal authenticated request.</summary>
        public async Task<ConnectionResult> TestConnectionAsync()
        {
            double since = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;
            var url = ApiUrl($"updates?since={since.ToString(System.Globalization.CultureInfo.InvariantCulture)}");
            if (url == null) return ConnectionResult.BadUrl();

            using var request = new HttpRequestMessage(HttpMethod.Get, url);
            request.Headers.TryAddWithoutValidation("authtoken", Token);

            try
            {
                using var response = await Http.SendAsync(request).ConfigureAwait(false);
                switch ((int)response.StatusCode)
                {
                    case 200: return ConnectionResult.Ok();
                    case 401:
                    case 403: return ConnectionResult.Unauthorized();
                    default: return ConnectionResult.Fail($"HTTP {(int)response.StatusCode}");
                }
            }
            catch (Exception ex)
            {
                return ConnectionResult.Fail(ex.Message);
            }
        }

        /// <summary>Short random record key (compact alphanumeric, like TimeTagger keys).</summary>
        public static string GenerateKey()
        {
            const string alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
            var chars = new char[10];
            for (int i = 0; i < chars.Length; i++)
                chars[i] = alphabet[Random.Shared.Next(alphabet.Length)];
            return new string(chars);
        }
    }
}

File: Taggd/TimeTaggerClient.swift
//
//  TimeTaggerClient.swift
//  Taggd
//
//  Thin client for a self-hosted TimeTagger backend.
//  https://github.com/almarklein/timetagger
//
//  Auth: the API token is sent in the `authtoken` HTTP header.
//  All endpoints live under <serverURL>/api/v2/ (enter the install root
//  as the server URL — for a subpath install use e.g. https://host/timetagger).
//

import Foundation

struct TimeTaggerClient {
    let serverURL: String
    let token: String

    /// Builds a client from the saved Server URL + API token, or nil if unconfigured.
    static func fromStoredSettings() -> TimeTaggerClient? {
        let defaults = UserDefaults.standard
        let url = defaults.string(forKey: "serverURL")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let token = defaults.string(forKey: "apiToken")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !url.isEmpty, !token.isEmpty else { return nil }
        return TimeTaggerClient(serverURL: url, token: token)
    }

    enum ConnectionResult: Equatable {
        case success(serverTime: Double)
        case unauthorized
        case badURL
        case failure(String)
    }

    /// Builds `<server>/api/v2/<endpoint>`, tolerating a trailing slash on the base.
    func apiURL(_ endpoint: String) -> URL? {
        var base = serverURL.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !base.isEmpty else { return nil }
        while base.hasSuffix("/") { base.removeLast() }
        return URL(string: base + "/api/v2/" + endpoint)
    }

    // MARK: - Records

    /// A fresh, compact alphanumeric record key (TimeTagger keys are short strings).
    static func generateKey() -> String {
        let alphabet = Array("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
        return String((0..<10).map { _ in alphabet.randomElement()! })
    }

    /// A TimeTagger record. A *running* record has `t1 == t2`; a finished one has `t2 > t1`.
    /// Tags are encoded inline in `ds` as `#tag` tokens.
    struct Record: Codable {
        let key: String
        let t1: Int
        let t2: Int
        let mt: Int
        let ds: String
    }

    enum PushResult: Equatable {
        case success
        case unauthorized
        case rejected(String)
        case badURL
        case failure(String)
    }

    enum RecordsResult {
        case success([Record])
        case unauthorized
        case badURL
        case failure(String)
    }

    /// Fetches records overlapping the `[start, end]` unix-second range. The server
    /// replies with `{"records": [...]}`; deleted records (ds prefixed "HIDDEN") are
    /// dropped so the caller only sees live entries.
    func fetchRecords(from start: Int, to end: Int) async -> RecordsResult {
        guard let url = apiURL("records?timerange=\(start)-\(end)") else { return .badURL }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue(token, forHTTPHeaderField: "authtoken")
        request.timeoutInterval = 15

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .failure("No response from server")
            }
            switch http.statusCode {
            case 200:
                struct Envelope: Decodable { let records: [Record] }
                let decoded = try JSONDecoder().decode(Envelope.self, from: data)
                let live = decoded.records.filter { !$0.ds.hasPrefix("HIDDEN") }
                return .success(live)
            case 401, 403:
                return .unauthorized
            default:
                return .failure("HTTP \(http.statusCode)")
            }
        } catch {
            return .failure(error.localizedDescription)
        }
    }

    /// PUTs one or more records. The body is a bare JSON array; the server replies
    /// with `{accepted, failed, errors}`.
    func pushRecords(_ records: [Record]) async -> PushResult {
        guard let url = apiURL("records") else { return .badURL }

        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue(token, forHTTPHeaderField: "authtoken")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 15
        do {
            request.httpBody = try JSONEncoder().encode(records)
        } catch {
            return .failure("Could not encode record")
        }

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .failure("No response from server")
            }
            switch http.statusCode {
            case 200:
                let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
                let failed = (object?["failed"] as? [Any]) ?? []
                if failed.isEmpty {
                    return .success
                }
                let errors = (object?["errors"] as? [String]) ?? []
                return .rejected(errors.first ?? "Server rejected the record")
            case 401, 403:
                return .unauthorized
            default:
                return .failure("HTTP \(http.statusCode)")
            }
        } catch {
            return .failure(error.localizedDescription)
        }
    }

    /// "Deletes" a record. TimeTagger keeps history, so deletion is a re-push of the
    /// same key with its description hidden; the app then filters `HIDDEN` records out.
    func deleteRecord(_ record: Record) async -> PushResult {
        let hidden = Record(
            key: record.key,
            t1: record.t1,
            t2: record.t2,
            mt: Int(Date().timeIntervalSince1970),
            ds: record.ds.hasPrefix("HIDDEN") ? record.ds : "HIDDEN " + record.ds
        )
        return await pushRecords([hidden])
    }

    /// Validates the server URL + token with a minimal authenticated request.
    /// `since=now` returns no records, so the payload stays tiny.
    func testConnection() async -> ConnectionResult {
        let since = Date().timeIntervalSince1970
        guard let url = apiURL("updates?since=\(since)") else { return .badURL }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue(token, forHTTPHeaderField: "authtoken")
        request.timeoutInterval = 15

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .failure("No response from server")
            }
            switch http.statusCode {
            case 200:
                let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
                let serverTime = (object?["server_time"] as? Double) ?? 0
                return .success(serverTime: serverTime)
            case 401, 403:
                return .unauthorized
            default:
                return .failure("HTTP \(http.statusCode)")
            }
        } catch {
            return .failure(error.localizedDescription)
        }
    }
}
